#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [32]:
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)
df.columns = df.columns.str.strip()
type(df)

pandas.core.frame.DataFrame

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [6]:
branch_profit = df.groupby('Branch')['gross income'].sum()

print(branch_profit)

most_profitable_branch = branch_profit.idxmax()
max_profit = branch_profit.max()

print(f"\nmost_profitable_branch: {most_profitable_branch} with_profit {max_profit:.2f}")

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

most_profitable_branch: C with_profit 5265.18


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [11]:
branch_profit = df.groupby('Branch')['gross income'].sum()

most_profitable_branch = branch_profit.idxmax()

city_for_branch = df.loc[df['Branch'] == most_profitable_branch, 'City'].unique()

print(f"most_profitable_branch : {most_profitable_branch}")
print(f"city: {city_for_branch}")

most_profitable_branch : C
city: ['Naypyitaw']


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [84]:
invoices_by_category = df.pivot_table(index='Branch', columns='Product line', values='Invoice ID', aggfunc='count' )
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000) 
print(invoices_by_category)

Product line  Electronic accessories  Fashion accessories  Food and beverages  Health and beauty  Home and lifestyle  Sports and travel
Branch                                                                                                                                 
A                                 60                   51                  58                 47                  65                 59
B                                 55                   62                  50                 53                  50                 62
C                                 55                   65                  66                 52                  45                 45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [22]:
max_branch = invoices_by_category['Electronic accessories'].idxmax()
max_value = invoices_by_category['Electronic accessories'].max()

print(f"max_invoice_branch': {max_branch} ({max_value} invoices)")

max_invoice_branch': A (60 invoices)


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Гіпотеза 1. 
В яких категоріях та філіях спостерігається найбільший валовий дохід? Чи є категорії, які приносять низький валовий дохід, але мають високі продажі?
Мета: Для оптимізації асортименту і планування інвестицій — фокусуватися на прибуткових категоріях і покращувати низькоприбуткові.

In [72]:
gross_income_by_category_branch = df.groupby(['Branch', 'Product line'])['gross income'].sum().unstack()
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000) 
gross_income_by_category_branch['Total'] = gross_income_by_category_branch.sum(axis=1)
gross_income_by_category_branch.loc['Total'] = gross_income_by_category_branch.sum(axis=0)


print("Валовий дохід за філіями та категоріями:")
print(gross_income_by_category_branch)


Валовий дохід за філіями та категоріями:
Product line  Electronic accessories  Fashion accessories  Food and beverages  Health and beauty  Home and lifestyle  Sports and travel       Total
Branch                                                                                                                                             
A                           872.2435             777.7385            817.2905            599.893           1067.4855           922.5095   5057.1605
B                           811.9735             781.5865            724.5185            951.460            835.6745           951.8190   5057.0320
C                           903.2845            1026.6700           1131.7550            791.206            661.6930           750.5680   5265.1765
Total                      2587.5015            2585.9950           2673.5640           2342.559           2564.8530          2624.8965  15379.3690


In [69]:
gross_income_by_category = df.groupby(['Branch', 'Product line'])['gross income'].sum()

sales_count_by_category = df.groupby(['Branch', 'Product line']).size()
summary = pd.DataFrame({
    'Gross Income': gross_income_by_category,
    'Sales Count': sales_count_by_category
})

summary['Income per Sale'] = summary['Gross Income'] / summary['Sales Count']
summary = summary.sort_values(by=['Branch', 'Gross Income'], ascending=[True, False])

print(summary)


                               Gross Income  Sales Count  Income per Sale
Branch Product line                                                      
A      Home and lifestyle         1067.4855           65        16.422854
       Sports and travel           922.5095           59        15.635754
       Electronic accessories      872.2435           60        14.537392
       Food and beverages          817.2905           58        14.091216
       Fashion accessories         777.7385           51        15.249775
       Health and beauty           599.8930           47        12.763681
B      Sports and travel           951.8190           62        15.351919
       Health and beauty           951.4600           53        17.952075
       Home and lifestyle          835.6745           50        16.713490
       Electronic accessories      811.9735           55        14.763155
       Fashion accessories         781.5865           62        12.606234
       Food and beverages          724

Аналіз
Найбільший валовий дохід в Філії С,  Філії А та В мають майже однаковий валовий дохід.
Найбільший валовий дохід по категорії має Foods and beverages
РОзглянемо окремо Філії та Категорії для формування рекомендацій для бізнесу.

1. Де найбільший валовий дохід?

Філія A:
Home and lifestyle — 1067.49
Sports and travel — 922.51

Філія B:
Sports and travel — 951.82
Health and beauty — 951.46
Найнижчий дохід на одиницю у категорії B Fashion accessories — 12.61, хоч кількість продажів висока. — 835.67

Філія C:
Food and beverages — 1131.76 (найвищий валовий дохід серед усіх)
Fashion accessories — 1026.67
Electronic accessories — 903.28

Філії А та В мають схожу структуру по категоріям з найвищим валовим доходом. Можна зробити висновок що дані філії схожа цільова аудиторія покупців.
Філія С взагалі в кошику категорій з найвищим валовим доходом має інші продукти

2. Проаналізуємо категорії з низьким валовим доходом, але виською кількістю продажів
Перевіряємо співвідношення валового доходу і кількості продажів:

У філії A Health and beauty — валовий дохід 599.89 (найнижчий у філії), але кількість продажів — 47 (найменша у філії, але не дуже сильно).
У філії B Fashion accessories — валовий дохід 781.59 при кількості продажів 62 — доволі багато продажів, але валовий дохід не найвищий.
У філії C Sports and travel — валовий дохід 750.57 при продажах лише 45 — низька кількість, але порівняно добрий дохід.

3. Аналіз доходу на одиницю продажу (маржинальність)
Найвищий дохід на продаж в Філії B = Health and beauty — 17.95, хоча валовий дохід не найвищий. Найнижчий дохід на одиницю в Філії B =  Fashion accessories — 12.61, хоч кількість продажів висока.
У філії C Food and beverages — дохід на продаж 17.15 — найвищий валовий дохід і висока маржинальність.  Категорія Home and lifestyle має найнижчу маржинальність, хоча кількість продажів не суттєво нижча інших категорій.
У Філії А найвища маржинальність — Home and lifestyle (16.42), і вона ж має найбільший валовий дохід.
Health and beauty — має найнижчу маржинальність (12.76) і найменший валовий дохід.

РЕКОМЕНДАЦІЇ

Фокус на філії C у категорії Food and beverages, оскільки тут найбільший валовий дохід і висока маржинальність.
Категорія Health and beauty у філії B має високу маржинальність, хоча загальний валовий дохід можна покращити — варто проаналізувати асортимент, підвищити продажі.
Fashion accessories у філії B — багато продажів, але низька маржинальність, можливо потрібно переглянути цінову політику та підвищити ціни, якщо це дозволяє моніторинг ринку та ціни конкурентів
У філії A варто підтримувати Home and lifestyle та Sports and travel — вони приносять найбільший дохід.
Категорії з нижчим доходом, але стабільними продажами (наприклад, Health and beauty у філії A), варто розглядати для маркетингових кампаній або оновлення асортименту.



Гіпотеза 2: Взаємозв’язок між кількістю одиниць у покупці (Quantity) та методом оплати (Payment)

In [75]:
avg_quantity_by_payment = df.groupby('Payment')['Quantity'].mean()

print("Avr_Quantity_payment:")
print(avg_quantity_by_payment)


Avr_Quantity_payment:
Payment
Cash           5.511628
Credit card    5.536977
Ewallet        5.484058
Name: Quantity, dtype: float64


Метою даного аналізу було виявити, чи існує залежність між методом оплати та кількістю одиниць товару в одному чеку. Зокрема, ми досліджували чи покупці, які обирають певну форму оплати (готівка, кредитна картка чи електронний гаманець), купують більше чи менше товарів у порівнянні з іншими.
Результати:
Середня кількість товарів у чеку за кожним методом оплати:
Кредитна картка: 5.54 одиниці
Готівка: 5.51 одиниці
Електронний гаманець: 5.48 одиниці
Отримані результати показують, що різниця між середніми значеннями не є суттєвою. Усі три методи оплати демонструють дуже схожу поведінку споживачів щодо кількості придбаних товарів.
Висновок:
Немає статистично або практично значущої різниці в кількості товарів у чеку залежно від методу оплати. Це дозволяє припустити, що вибір форми оплати не впливає на обсяг покупки. Іншими словами, попит на кожен метод оплати є збалансованим і всі вони однаково використовуються незалежно від кількості одиниць товару в чеку.

Гипотеза 3. Вплив  статі на середній чек  і вибір категорій товарів .

In [83]:
avg_total_by_productline_gender = df.groupby(['Product line', 'Gender'])['Total'].mean().reset_index()

sorted_avg = avg_total_by_productline_gender.sort_values(by=['Gender', 'Total'], ascending=[True, False])

print("Avg_total_by_product_line_gender:")
print(f"{'Product line':<25} {'Gender':<10} {'Total':>10}")
for _, row in sorted_avg.iterrows():
    print(f"{row['Product line']:<25} {row['Gender']:<10} {row['Total']:>10.2f}")

Avg_total_by_product_line_gender:
Product line              Gender          Total
Home and lifestyle        Female         380.21
Food and beverages        Female         368.57
Sports and travel         Female         324.71
Electronic accessories    Female         322.64
Fashion accessories       Female         317.06
Health and beauty         Female         290.02
Health and beauty         Male           348.10
Sports and travel         Male           340.36
Electronic accessories    Male           316.69
Home and lifestyle        Male           294.14
Fashion accessories       Male           291.08
Food and beverages        Male           273.50


Аналіз
Жінки витрачають більше в кожній категорії товарів, ніж чоловіки:
Найбільший середній чек серед жінок — "Home and lifestyle" (380.21).
Найнижчий — "Health and beauty" (290.02).
У чоловіків найвищий чек — "Health and beauty" (348.10), а найнижчий — "Food and beverages" (273.50).
Категорія "Health and beauty" має цікаву поведінку:
У жінок це найменший середній чек, а в чоловіків — найвищий.
Це може свідчити про те, що жінки купують частіше, але меншими чеками, тоді як чоловіки — рідше, але дорожчі одиниці (наприклад, парфуми, набори по догляду).
Категорія "Food and beverages" має найнижчі середні чеки серед чоловіків і другий за величиною серед жінок: можливо, жінки частіше купують їжу для сім’ї та великі об'єми, а чоловіки — для себе (дрібні покупки).
Категорії "Home and lifestyle" та "Sports and travel":
У жінок середній чек вищий, особливо помітна різниця в "Home and lifestyle"

РЕКОМЕНДАЦІЇ ДЛЯ БІЗНЕСУ
1. Цільовий маркетинг за статтю
Для жінок:
Пропонувати акції чи пропозиції у категоріях: Home and lifestyle, Food and beverages, Fashion accessories.
Рекламні кампанії з акцентом на домашній комфорт, стиль і гастрономію.

Для чоловіків:
Фокус на "Health and beauty" та "Sports and travel".
Створити окремі промо на дорогі косметичні/брендові продукти.

2. Персоналізація пропозицій
Враховувати стать при створенні рекомендацій у мобільному застосунку або email-розсилці з врахуванням витрат по категоріям



